File: src/e2e/rtmp_passthrough_test.go
//go:build e2e

package e2e

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"testing"
	"time"

	"Theatrum/constants"
)

func TestRTMP_PassthroughPipeline(t *testing.T) {
	liveStreamKey := "e2e-secret-key"
	username := "alice"
	token := computeXORToken(liveStreamKey, username)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/user/{username}":
    stream:
      type: live
      path: "live/{username}"
      live_stream_key: "%s"
      auth_token_template: "{username}"
      distribution:
        hls:
          segment_duration: 2
          window_size: 3
`, liveStreamKey),
	})

	// Start RTMP stream
	rtmpURL := fmt.Sprintf("rtmp://127.0.0.1:%d/user/%s", ts.RTMPPort, username)
	cmd := streamToRTMP(t, rtmpURL, token, 15)

	// Wait for the playlist to appear on disk
	// Passthrough streams go into {path}/default/
	playlistPath := filepath.Join(ts.DataDir, "live", username, constants.DefaultQuality, constants.SubPlaylist)
	waitForFile(t, playlistPath, 15*time.Second)

	// Wait for at least one segment
	segmentPath := filepath.Join(ts.DataDir, "live", username, constants.DefaultQuality, "segment_000.ts")
	waitForFile(t, segmentPath, 15*time.Second)

	// Fetch the sub playlist via HTTP
	subPlaylistURL := fmt.Sprintf("http://%s/user/%s/%s/%s", ts.HTTPAddr, username, constants.DefaultQuality, constants.SubPlaylist)
	resp := waitForHTTP(t, subPlaylistURL, 15*time.Second)
	resp.Body.Close()

	// Verify it's a live playlist (no #EXT-X-ENDLIST)
	status, body := httpGetBody(t, subPlaylistURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for sub playlist, got %d", status)
	}
	m3u8 := parseM3U8(body)
	if m3u8.IsVOD {
		t.Error("Expected live playlist (no #EXT-X-ENDLIST)")
	}
	if m3u8.SegmentCount == 0 {
		t.Error("Expected at least one segment in playlist")
	}

	// Stop the FFmpeg stream
	cmd.Process.Kill()
	cmd.Wait()

	// Wait for cleanup (reconnect_delay=1 + cleanup_delay=1)
	// The stream files should be deleted after cleanup
	time.Sleep(4 * time.Second)

	// Verify cleanup: the stream directory or files should be removed
	waitForFileAbsent(t, segmentPath, 10*time.Second)
}

func TestRTMP_PassthroughServesSegmentsViaHTTP(t *testing.T) {
	liveStreamKey := "e2e-secret-key2"
	username := "bob"
	token := computeXORToken(liveStreamKey, username)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/user/{username}":
    stream:
      type: live
      path: "live/{username}"
      live_stream_key: "%s"
      auth_token_template: "{username}"
      distribution:
        hls:
          segment_duration: 2
          window_size: 3
`, liveStreamKey),
	})

	// Start RTMP stream
	rtmpURL := fmt.Sprintf("rtmp://127.0.0.1:%d/user/%s", ts.RTMPPort, username)
	streamToRTMP(t, rtmpURL, token, 12)

	// Wait for segments to appear
	segmentPath := filepath.Join(ts.DataDir, "live", username, constants.DefaultQuality, "segment_000.ts")
	waitForFile(t, segmentPath, 15*time.Second)

	// Fetch a segment via HTTP
	segURL := fmt.Sprintf("http://%s/user/%s/%s/segment_000.ts", ts.HTTPAddr, username, constants.DefaultQuality)
	resp := waitForHTTP(t, segURL, 10*time.Second)
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("Expected 200 for segment, got %d", resp.StatusCode)
	}

	// Verify segment has content
	stat, err := os.Stat(segmentPath)
	if err != nil {
		t.Fatalf("Failed to stat segment: %v", err)
	}
	if stat.Size() == 0 {
		t.Error("Segment file is empty")
	}
}

File: src/test/java/com/fjourdren/theatrum/e2e/RtmpPassthroughTest.java
package com.fjourdren.theatrum.e2e;

import com.fjourdren.theatrum.domain.constant.VideoConstants;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Port of Go's {@code rtmp_passthrough_test.go}: a real FFmpeg publisher pushing the test video
 * over RTMP, transmuxed to HLS with codec copy and served over HTTP.
 */
class RtmpPassthroughTest {

    private static final String KEY = "e2e-secret-key";

    private static final String CHANNELS = """
              "/user/{username}":
                stream:
                  type: live
                  path: "live/{username}"
                  live_stream_key: "%s"
                  auth_token_template: "{username}"
                  distribution:
                    hls:
                      segment_duration: 2
                      window_size: 3
            """.formatted(KEY);

    @TempDir
    static Path tempDir;

    static TestServerSupport server;

    @BeforeAll
    static void startServer() {
        TestServerSupport.requireFfmpeg();
        server = TestServerSupport.start(tempDir, CHANNELS);
    }

    @AfterAll
    static void stopServer() {
        if (server != null) {
            server.close();
        }
    }

    private Path defaultQualityDir(String username) {
        return server.dataDir.resolve("live").resolve(username).resolve(VideoConstants.DEFAULT_QUALITY);
    }

    @Test
    void transmuxesToALivePlaylistAndCleansUpWhenThePublisherLeaves() {
        String username = "alice";
        Process publisher = server.publish("user/" + username, TestServerSupport.xorToken(KEY, username), 12);

        // Passthrough streams land in {path}/default/.
        Path segment = defaultQualityDir(username).resolve("segment_000.ts");
        TestServerSupport.waitForFile(defaultQualityDir(username).resolve(VideoConstants.SUB_PLAYLIST),
                Duration.ofSeconds(20));
        TestServerSupport.waitForFile(segment, Duration.ofSeconds(20));

        String playlistUrl = "/user/%s/%s/%s".formatted(username, VideoConstants.DEFAULT_QUALITY,
                VideoConstants.SUB_PLAYLIST);
        server.waitForHttp(playlistUrl, Duration.ofSeconds(20));

        var response = server.get(playlistUrl);
        assertThat(response.status()).isEqualTo(200);
        assertThat(response.header("Cache-Control")).isEqualTo("no-cache, no-store, must-revalidate");

        var playlist = TestServerSupport.parseM3u8(response.body());
        assertThat(playlist.vod()).as("a live playlist has no #EXT-X-ENDLIST").isFalse();
        assertThat(playlist.segmentCount()).isPositive();

        TestServerSupport.kill(publisher);

        // reconnect_delay (1s) then cleanup_delay (1s) before the files go away.
        TestServerSupport.waitForFileAbsent(segment, Duration.ofSeconds(20));
    }

    @Test
    void servesSegmentsOverHttp() throws Exception {
        String username = "bob";
        Process publisher = server.publish("user/" + username, TestServerSupport.xorToken(KEY, username), 10);

        Path segment = defaultQualityDir(username).resolve("segment_000.ts");
        TestServerSupport.waitForFile(segment, Duration.ofSeconds(20));

        String segmentUrl = "/user/%s/%s/segment_000.ts".formatted(username, VideoConstants.DEFAULT_QUALITY);
        var response = server.waitForHttp(segmentUrl, Duration.ofSeconds(15));
        assertThat(response.status()).isEqualTo(200);
        assertThat(response.header("Content-Type")).isEqualTo("video/mp2t");

        assertThat(java.nio.file.Files.size(segment)).isPositive();
        assertThat(server.getBytes(segmentUrl)).isNotEmpty();

        TestServerSupport.kill(publisher);
    }
}
